engine/src/main/scala/cromwell/engine/workflow/tokens/JobExecutionTokenDispenserActor.scala

```diff
@@ -0,0 +1,130 @@
+package cromwell.engine.workflow.tokens
+
+import akka.actor.{Actor, ActorLogging, ActorRef, Props, Terminated}
+import cromwell.core.JobExecutionToken
+import JobExecutionToken._
+import cromwell.engine.workflow.tokens.JobExecutionTokenDispenserActor._
+import cromwell.engine.workflow.tokens.TokenPool.TokenPoolPop
+
+import scala.collection.immutable.Queue
+import scala.language.postfixOps
+
+class JobExecutionTokenDispenserActor extends Actor with ActorLogging {
+
+  /**
+    * Lazily created token pool. We only create a pool for a token type when we need it
+    */
+  var tokenPools: Map[JobExecutionTokenType, TokenPool] = Map.empty
+  var tokenAssignments: Map[ActorRef, JobExecutionToken] = Map.empty
+
+  override def receive: Actor.Receive = {
+    case JobExecutionTokenRequest(tokenType) => sendTokenRequestResult(sender, tokenType)
+    case JobExecutionTokenReturn(token) => unassign(sender, token)
+    case Terminated(terminee) => onTerminate(terminee)
+  }
+
+  private def sendTokenRequestResult(sndr: ActorRef, tokenType: JobExecutionTokenType): Unit = {
+    if(tokenAssignments.contains(sndr)) {
+      sndr ! JobExecutionTokenDispensed(tokenAssignments(sndr))
+    } else {
+      context.watch(sndr)
+      val updatedTokenPool = getTokenPool(tokenType).pop() match {
+        case TokenPoolPop(newTokenPool, Some(token)) =>
+          assignAndSendToken(sndr, token)
+          newTokenPool
+        case TokenPoolPop(sizedTokenPoolAndQueue: SizedTokenPoolAndActorQueue, None) =>
+          val (poolWithActorEnqueued, positionInQueue) = sizedTokenPoolAndQueue.enqueue(sndr)
+          sndr ! JobExecutionTokenDenied(positionInQueue)
+          poolWithActorEnqueued
```

Comment:
for completeness (even if the other type of pool is Infinite so it won't happen), the 
`case TokenPoolPop(somethingNotASizedTokenPoolAndActorQueue, None)` is not caught here